Source/Dafny/Compilers/Compiler-python.cs

```diff
@@ -325,7 +331,7 @@ protected class ClassWriter : IClassWriter {
 
     protected override void EmitLiteralExpr(ConcreteSyntaxTree wr, LiteralExpr e) {
       if (e.Value is bool value) {
-        wr.Write(value ? "True" : "False");
+        wr.Write("helpers.printLiteralExpr({0})", value);
```

Comment:
I'm not sure I understand the goal of this. Before, you were printing a python boolean, and now you are writing a function that returns a string true/false instead? It seems you are changing the semantics.

Reply:
So, as the original code is written in dafny the output should be as what dafny code is compiled not python code, that is why I have written a runtime helper function which takes in python's input and prints out as per dafny.

Reply:
I already understood that point. What I don't understand is that `EmitLiteralExpr` literally means "print the python equivalent of this literal expression", which is `True/False` (without double quotes).
The discrepancy came from python's print function, that's the function you need to provide a helper for, not the literal values of true or false.

The code you write will fail to compile the following example:
```
if false {
  print 1;
} else {
  print 0;
}
```

because your code will render the python equivalent of:

```
if helpers.printLiteralExpr(false) {
  print 1;
} else {
  print 0;
}
```

the value that `helpers.printLiteralExpr(false)` will return is `"false"`, which is a non-empty string that is coerced to bool, so 1 will be displayed instead of 0.

Reply:
I am afraid that's not the case, this is will only be called if the statement is a ```print``` statement and the value in that is a ```literal value```, It will not replace all the ```literal values```.

Reply:
The function `TrExpr` is responsible for calling `EmitLiteralExpr`, and `TrExpr` is virtually called from 182 different places. I have serious doubts that these 182 places are all for the case when we are in a `print statement`.